src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMutation.cs

```diff
@@ -0,0 +1,255 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.IR;
+using Microsoft.PowerFx.Functions;
+using Microsoft.PowerFx.Interpreter;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Texl.Builtins
+{
+    // Patch(dataSource:*[], Record, Updates1, Updates2,…)
+    internal class PatchImpl : PatchFunction, IAsyncTexlFunction5
+    {
+        public async Task<FormulaValue> InvokeAsync(IServiceProvider runtimeServiceProvider, FormulaType irContext, FormulaValue[] args, CancellationToken cancellationToken)
+        {
+            MutationProcessUtils.ThrowIfPFxV1NotActive(runtimeServiceProvider.GetService<Features>(), "Patch");
+
+            var arg0 = args[0];
+
+            if (args[0] is LambdaFormulaValue arg0lazy)
+            {
+                arg0 = await arg0lazy.EvalAsync().ConfigureAwait(false);
+            }
+
+            if (arg0 is not TableValue tableValue)
+            {
+                return arg0;
+            }
+
+            if (args[1] is not RecordValue baseRecord)
+            {
+                return args[1];
+            }
+
+            var dvalue = MutationUtils.MergeRecords(args.Skip(2));
+
+            if (!dvalue.IsValue)
+            {
+                return dvalue.ToFormulaValue();
+            }
+
+            var result = await tableValue.PatchAsync(baseRecord, dvalue.Value, cancellationToken).ConfigureAwait(false);
+
+            // If the base record is not found, then append update record.
+            if (result.IsError && result.Error is ErrorValue errorvalue && errorvalue.Errors.Any(err => err.Kind == ErrorKind.NotFound))
+            {
+                return (await tableValue.AppendAsync(dvalue.Value, cancellationToken).ConfigureAwait(false)).ToFormulaValue();
+            }
+
+            return result.ToFormulaValue();
+        }
+    }
+
+    // If arg1 is pure PFx record, it will return a runtime internal error.
```

Comment:
>runtime internal error [](http://example.com/codeflow?start=52&length=22)

It should not be an internal error (those are bugs), it should be an invalid argument.